### PR TITLE
Key events

### LINKED ISSUE #55: Key events
```py
# Syntax option 1
if key_held['a'] and key_held['b']:

end
```

AC:
 * [x] Add a user-definable `key_pressed()` function
    * If no key functions are defined, the event handlers should *not* be created, as they will slow down the program unnecessarily
 * [x] Filter out repeated key pressed
    * The user-defined `key_pressed` should only be called once when the key goes from not pressed to pressed
 * [x] Add a user-definable `key_released()` function
    * The user-defined `key_released` should only be called once when the key goes from pressed to not pressed
 * [x] OPTIONAL - add a user-definable `key_repeated()` function
    * The user-defined `key_repeated` should be called each time the browser re-issues the key pressed event, which will occur when the key is held
 * [x] Add `key_held` dict to detect if a key is held, necessary for diagonal movement (see code sample above)
    * Should be a global immutable dictionary, for each key the value is whether that key is currently pressed or not
 * [x] When a special key (e.g. arrow key) is pressed, make it detectable by `key` (unlike Processing that uses `keyCode` to distinguish)
 * [x] Unit tests
   * [x] Must include a "game" example where a player can travel diagonally by pressing 2 keys simultaneously
 * [x] Assign @Descent098 to this ticket when ready to start documentation and communicate how it works
 * [x] Document regular key presses
 * [x] Document special keys
 * [x] Document key pressed events

### PR DESCRIPTION
**Pull request by lawruble13**
Got key events working using `ipyevents`

### CODE REVIEW

**Comment by lawruble13** [comment]
With this merge, keyboard events should *work*, but most of the requested additional functionality has not been added.

**Comment by lawruble13** [comment]
The most recent commit:

- Adds the user-definable `key_pressed` function, which will be called when any key is pressed
- Adds the user-definable `key_repeated` function, which will be called when a key is held, and the key press event is repeated
- Adds the user-definable `key_released` function, which will be called when any key is pressed
- Adds the global immutable `key`, which holds the most recent key for which an event is triggered.
    - All events above should run with `key` being the key which triggered that event, but there may be a race condition if two or more keyboard events occur simultaneously. Testing required.
- Adds the `keys_held` function, instead of a dict as listed in the issue. This is because we cannot make the dict immutable.
    - This deviates from the specified syntax, allowing calls of the form `keys_held("a", "d")`, as well as specifying if certain keys should be held or not, e.g. `keys_held("a", "d", pattern=[True, False])`. In a game, this allows a check for forward/upward movement to take the form `keys_held("ArrowUp", "ArrowDown", [True, False])`
- Special keys can be detected using the `key` variable, and their status can be checked using `keys_held`.
    - Some special keys (e.g. up arrow, down arrow, spacebar, F12) have special meanings captured by the browser. Supposedly, some of these issues are fixed in ipyevents 0.4.0 (on npm, 1.4.0), but I am still experiencing issues
- One test has been added, which incorporates pressing, releasing, repeating, and draw-loop functionality. It's not a "game" as such, but does include diagonal movement
- The rate of keyboard events has been limited to one per `1000//FRAME_RATE` ms, to reduce lag caused by too many events

**Comment by lawruble13** [comment]
Also note that this most recent commit changes most returns of the main loop to use the `stop` function, as the `stop` function also disables keyboard listeners. Prior to this change, keyboard listeners accumulated, and could start to cause lag as multiple drawings each listened to every keyboard event.

**Comment by lawruble13** [comment]
Keeping track of what is currently pressed is causing some issues, partially resolved in upcoming commit. Namely, issues are encountered with any duplicated keys, and keys which produce different output depending on Shift and/or Caps Lock. As there are two shift keys, Shift may be truly pressed twice sequentially, without an intervening release. Additionally, if the A key is pressed, then Shift xor Caps Lock is pressed, then A is released, then we see "a" pressed, "Shift"/"CapsLock" pressed, "A" released (note "a" and "A" are different *keys*, with the same *code*). This has been partially resolved by clearing both options for all codes which may produce different keys when Shift or Caps Lock are active. This is, at present, layout-specific (e.g. UK keyboards with £ on 3 have issues, as do I on my standard mapping, Dvorak).

File: test/KeyEventsUnits.ipynb [hunk 1]
```diff
@@ -0,0 +1,1088 @@
...
+    "    text(f\"{repeated_count} times\", 50, 250)\n",
+    "\n",
+    "def key_pressed():\n",
+    "    global repeated_key, repeated_count\n",
+    "    repeated_key = key\n",
+    "    repeated_count = 0\n",
+    "    \n",
+    "def key_repeated():\n",
+    "    global repeated_key, repeated_count\n",
+    "    repeated_key = key\n",
```
Comment (AlphaRLee) [inline test/KeyEventsUnits.ipynb:727]: Small detail - your `key_repeated` test is missing the `<space>` output

**Comment by AlphaRLee** [comment]
When running the arrow key movement test, I hit the common issue of having the browser scroll up and down when pressing up and down. Is there an easy way for the user or the browser to prevent this behaviour? My brief glance at [this StackOverflow](https://stackoverflow.com/questions/8916620/disable-arrow-key-scrolling-in-users-browser/8916697) page says that the Javascript event needs to be cancelled; not sure if we can access that in Python.

image

**Comment by lawruble13** [comment]
As I mentioned in a previous comment: 
> * Special keys can be detected using the `key` variable, and their status can be checked using `keys_held`.
>   
>   * Some special keys (e.g. up arrow, down arrow, spacebar, F12) have special meanings captured by the browser. Supposedly, some of these issues are fixed in ipyevents 0.4.0 (on npm, 1.4.0), but I am still experiencing issues. 

We don't explicitly have access to the JS (Typescript), unless we sculpt it all ourselves. Apparently, Jupyter Lab 2.0 broke a lot of things, and this may be one of them. I may take a look at the ipyevents library and see if I can get it working, and if so set up an upstream pr.

**Comment by AlphaRLee** [comment]
Thanks, doesn't sound urgent now, so we'll address those issues later.

**Comment by Descent098** [comment]
So here's what I have for the docs, let me know if I missed any functions/vars. I will submit the docs changes in a seperate branch and comment here with it:

## Global vars
key->str; the last key pressed

## Methods
keys_held(x:str); a function that returns true if x is held

## User definable functions
key_pressed(); user definable function that is procced when a key input is given
key_released(); user definable function that is procced when a key that has been pressed is released
key_repeated(); a user definable function that is procced when a key is held down

**Comment by lawruble13** [comment]
Those all look to be correct, and I don't think anything's missing!

**Comment by lawruble13** [comment]
I've found the issue in the ipyevents library, made a quick fix, and opened a PR for it over there. In the event the issue is intended behaviour, I do have a fix for the arrow keys, but it involves some trade-offs. The issue is that currently, keyboard events are throttled to only once every 33 ms, but the library doesn't stop the propagation of events if they're throttled, so holding the up or down key causes scrolling. The fix is to remove the throttling, which causes significant lag when any keys are held 😐 . If you want to test that, you can remove the `wait=1000 // FRAME_RATE` keyword argument on line 47 of core.py

File: test/KeyEventsUnits.ipynb (hunk 1 above)
Comment (lawruble13) [reply test/KeyEventsUnits.ipynb:727]: Fixed, and reworked the other examples for more efficiency in the event handler

**Comment by Descent098** [comment]
Docs for all the features done #62 (since dev doesn't have the current docs changes it says there's way more files changed than necessary)

**Comment by AlphaRLee** [comment]
We can just live with the bug until it's resolved, I think if we make all keys lag then we would lose the value from keypressing. Thanks for making the PR on the ipyevents library, that's amazing you went that far! 😄

**Review by AlphaRLee — APPROVED** [review]
Looks good to me. I didn't thoroughly review anything other than the `ignite_global` flag since I assume a majority of it is copy-pasted and tested

**Review by Descent098 — APPROVED** [review]
> Looks good to me. I didn't thoroughly review anything other than the `ignite_global` flag since I assume a majority of it is copy-pasted and tested

@AlphaRLee I think this review is for the tech debt branch #56 

But I will also give an approval to this branch since I tested it during writing the documentation.

**Pull request merged**